contributors/design-proposals/storage/attacher-detacher-refactor-for-local-storage.md

```diff
@@ -0,0 +1,264 @@
+    * [Volume plugin change](#volume-plugin-change)
+* [Future](#future)
+
+## Summary
+
+Today, the workflow for a volume to be used by pod is:
+- attach a remote volume to the node instance (if it is attachable)
+- wait for the volume to be attached (if it is attachable)
+- mount the device to a global path (if it is attachable)
+- mount the global path to a pod directory
+
+It is ok for remote block storage plugins which have a remote attach api,such as `GCE PD`, `AWS EBS`
+and remote fs storage plugins such as `NFS`, and `Cephfs`.
+But it is not so good for plugins which need local attach such as `fc`, `iscsi` and `RBD`.
+It is not so good for local storage neither which is not attachable but needs `MountDevice`
+
+## Motivation
+
+### Goals
+
+ Update Attacher/Detacher interfaces for local storage
+
+### Non-Goals
+
+ Update `fc`, `iscsi` and `RBD` implementation according to the new interfaces
+
+## Proposal
+
+Here we propose to only update the Attacher/Detacher interfaces for local storage.
+We may expand it in future to `iscsi`, `RBD` and `fc`, if we figure out how to prevent multiple local attach without implementing attacher interface.
+
+## Implementation
+
+### Volume plugin interface change
+
+We can create a new interface `DeviceMounter`,  move `GetDeviceMountPath` and `GetDeviceMountPath` from `Attacher`to it.
+
+Any one who implement the `Attacher` interface must implement `DeviceMounter` interface.
+```
+// Attacher can attach a volume to a node.
+type Attacher interface {
+        // Attaches the volume specified by the given spec to the node with the given Name.
+	// On success, returns the device path where the device was attached on the
+	// node.
+	Attach(spec *Spec, nodeName types.NodeName) (string, error)
+
+	// VolumesAreAttached checks whether the list of volumes still attached to the specified
+	// node. It returns a map which maps from the volume spec to the checking result.
+	// If an error is occurred during checking, the error will be returned
+	VolumesAreAttached(specs []*Spec, nodeName types.NodeName) (map[*Spec]bool, error)
+
+	// WaitForAttach blocks until the device is attached to this
+	// node. If it successfully attaches, the path to the device
+	// is returned. Otherwise, if the device does not attach after
+	// the given timeout period, an error will be returned.
+	WaitForAttach(spec *Spec, devicePath string, pod *v1.Pod, timeout time.Duration) (string, error)
+}
+
+// DeviceMounter can mount a block volume to a global path.
+type DeviceMounter interface {
+	// GetDeviceMountPath returns a path where the device should
+	// be mounted after it is attached. This is a global mount
+	// point which should be bind mounted for individual volumes.
+	GetDeviceMountPath(spec *Spec) (string, error)
+
+	// MountDevice mounts the disk to a global path which
+	// individual pods can then bind mount
+	MountDevice(spec *Spec, devicePath string, deviceMountPath string) error
```

Comment:
It would be worth noting that `devicePath` comes from `Attach` or `WaitForAttach` return values and can be empty if a plugin does not implement any of these calls.

Reply:
updated, thanks